### PR TITLE
Basic to mpoly

### PR DESCRIPTION
**Pull request by srajangarg**
Major WIP

### CODE REVIEW

**Comment by srajangarg** [comment]
MExprPoly remains. Suggestions for more testcases required.

File: symengine/polys/basic_conversions.h [hunk 1]
```diff
@@ -10,14 +10,14 @@ namespace SymEngine
 // using `gen` as the genarator. Throws, if poly constructions not possible.
 // `ex` is the optional parameter for expanding the given `basic` or not.
 template <typename P>
-RCP<const P> from_basic(const RCP<const Basic> &basic,
-                        const RCP<const Basic> &gen, bool ex = false);
+RCP<const P> upoly_from_basic(const RCP<const Basic> &basic,
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Can't this be the same function for both univariates and multivariates?

File: symengine/polys/basic_conversions.h [hunk 2]
```diff
@@ -229,6 +228,217 @@ RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex)
...
+        } else {
+            dict_set(zero_v, x);
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        Dict res = apply(*x.coef_, gens);
+        for (auto const &it : x.dict_)
+            res += apply(*it.first, gens) * apply(*it.second, gens);
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Using `apply` here would build `gens_pow` again right?

File: symengine/polys/basic_conversions.h [hunk 3]
```diff
@@ -229,6 +228,217 @@ RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex)
...
+                found = false;
+
+                for (auto powr : ite->second) {
+                    tmp = div(it, powr);
+                    if (is_a<const Integer>(*tmp)) {
+                        i = rcp_static_cast<const Integer>(tmp);
+                        if (i->is_positive()) {
+                            zero_v[gens_map[pow(ite->first, powr)]]
+                                = i->as_int();
+                            found = true;
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Don't you need to set `coef = one` here?

File: symengine/polys/basic_conversions.h (hunk 1 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: No, I tried. Internally, they have different operations but their signatures match. Maybe I can do it using metaprogramming.

File: symengine/polys/basic_conversions.h (hunk 2 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: Yes, I plan to add a boolean, to check if it's already constructed or not.

Comment (isuruf) [inline symengine/polys/basic_conversions.h]: You can also create `gens_pow` in the constructor.

File: symengine/polys/basic_conversions.h (hunk 1 above)
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: You can do this using templates. Let me know if you need any help.
What do you mean by metaprogramming?

File: symengine/polys/basic_conversions.h (hunk 3 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: No, consider `2**(x + y)` where the only generator is `2**y`. At `x` the coeff will be updated to `2**x`, and at `y`, we don't want to make the coeff = 1. (It should throw, for MIntPolys)

File: symengine/polys/basic_conversions.h (hunk 1 above)
Comment (srajangarg) [inline symengine/polys/basic_conversions.h]: I tried this (and `MSymEnginePoly` for the multivariate version, instead of `UPolyBase`)
```
template <typename T, typename P>
enable_if_t<std::is_base_of<UPolyBase<T, P>, P>::value, RCP<const P>>
from_basic(..)
```
but then, wouldn't this become a 2 template parameter functions? It's not able to deduce the "container" automatically

File: symengine/tests/polynomial/test_basic_conversions.cpp [hunk 4]
```diff
@@ -290,14 +292,6 @@ TEST_CASE("basic_to_poly UInt", "[b2poly]")
...
     gen = pow(i9, x);
     CHECK_THROWS_AS(from_basic<UIntPoly>(basic, gen), SymEngineException);
-
-    // x + y
-    basic = add(x, y);
-    CHECK_THROWS_AS(from_basic<UIntPoly>(basic), SymEngineException);
-
-    // x + 1/x
-    basic = add(x, div(one, x));
-    CHECK_THROWS_AS(from_basic<UIntPoly>(basic), SymEngineException);
```
Comment (isuruf) [inline symengine/tests/polynomial/test_basic_conversions.cpp]: Why is this removed?

Comment (srajangarg) [inline symengine/tests/polynomial/test_basic_conversions.cpp]: Duplicates. They exist right above too.

File: symengine/polys/msymenginepoly.cpp [hunk 5]
```diff
@@ -123,6 +123,8 @@ unsigned int reconcile(vec_uint &v1, vec_uint &v2, set_basic &s,
         }
         pos++;
     }
+
+    // SYMENGINE_ASSERT(v1.size)
```
Comment (isuruf) [inline symengine/polys/msymenginepoly.cpp]: Can you remove this comment ?

**Comment by isuruf** [comment]
Are you going to add `MExprPoly` here or in another PR?

File: symengine/tests/polynomial/test_basic_conversions.cpp [hunk 6]
```diff
@@ -466,3 +460,107 @@ TEST_CASE("basic_to_poly UIntFlint", "[b2poly]")
...
+    RCP<const Symbol> z = symbol("z");
+    RCP<const Basic> xb2 = div(x, i2);
+    RCP<const Basic> twopx = pow(i2, x);
+    RCP<const Basic> twopy = pow(i2, y);
+    RCP<const MIntPoly> poly1, poly2, poly3, poly4, poly5, poly6, poly7, poly8;
+
+    // x + y
+    basic = add(x, y);
+    gens = {x, y};
+    poly1 = from_basic<MIntPoly>(basic, gens);
```
Comment (isuruf) [inline symengine/tests/polynomial/test_basic_conversions.cpp]: Can you add a test case without giving `gens` explicitly to test that it works?

**Comment by srajangarg** [comment]
I think I will add MExprPoly in this PR itself (should be a small addition). Writing tests for it will take time.

File: symengine/tests/polynomial/test_basic_conversions.cpp (hunk 6 above)
Comment (srajangarg) [inline symengine/tests/polynomial/test_basic_conversions.cpp]: Done

**Comment by srajangarg** [comment]
@isuruf please have a look
